lib/base/CMakeLists.txt

```diff
@@ -0,0 +1,29 @@
+# -----------------------------------------------------------------------------
+# sources
+# -----------------------------------------------------------------------------
+target_sources(${PROJECT_NAME} 
+    PRIVATE
+        debugCodes.cpp
+)
+
+set(headers
+    api.h
+    debugCodes.h
+)
+
+# -----------------------------------------------------------------------------
+# promoted headers
+# -----------------------------------------------------------------------------
+mayaUsd_promoteMayaUsdHeader()
+mayaUsd_promoteHeaderList(HEADERS ${headers} SUBDIR base)
+
+# -----------------------------------------------------------------------------
+# install
+# -----------------------------------------------------------------------------
+install(FILES ${headers}
+    DESTINATION ${CMAKE_INSTALL_PREFIX}/include/mayaUsd/base
+)
+
+install(FILES ${CMAKE_BINARY_DIR}/include/mayaUsd/mayaUsd.h
+    DESTINATION ${CMAKE_INSTALL_PREFIX}/include/mayaUsd
+)
```

Comment:
I may be missing something here too, but promoting **and** installing feels somewhat redundant to me, and I'm not crazy about how the promote function injects `#pragma` directives on the fly. Is there a way to collapse this? Or maybe that's also planned for a later pass?

Reply:
Promoting and installing are kind of orthogonal to each other. The problem with collapsing them is that "install stage" doesn't necessarily need to happen all the time but the "build stage" does. 

Also relying on install stage to access the header files doesn't really work. What happens with incremental builds? That's why we include headers from the build directory when building the project. Also when the incremental build happens ONLY the headers files that have touched are promoted.

Please let me know if this answers your question or perhaps I got your question confused with something else.

Reply:
As I was reading your above comment in `lib/CMakeLists.txt `, I realized that having `${CMAKE_BINARY_DIR}/include` in the PUBLIC list makes it a bit vague when other projects link against mayaUsd project. 

This is now added in the PRIVATE entry list and projects that links against mayaUsd now need to explicitly add `${CMAKE_BINARY_DIR}/include` in their `target_include_directories`. 

mayaUsd 
```
target_include_directories(${PROJECT_NAME} 
    PUBLIC
        ${MAYA_INCLUDE_DIRS}
        ${PXR_INCLUDE_DIRS}
        $<$<BOOL:${UFE_FOUND}>:${UFE_INCLUDE_DIR}>
    PRIVATE
        ${CMAKE_BINARY_DIR}/include
        ${CMAKE_CURRENT_SOURCE_DIR}
)
```

_mayaUsd 
```
add_library(${PYTHON_TARGET_NAME} SHARED)
target_include_directories(${PYTHON_TARGET_NAME}
    PRIVATE
        ${CMAKE_BINARY_DIR}/include
)
target_link_libraries(${PYTHON_TARGET_NAME}
    PRIVATE
        mayaUsd
)
```

please see 904b394d6ff38

Reply:
I may again be exposing my CMake ignorance, but is this not what the INTERFACE keyword is for?

It seems like things that link against mayaUsd shouldn't need to do anything special (like add an include directory) other than just declare their dependence on mayaUsd. The build and install requirements necessary for a dependency of a downstream library should all be determined by the properties of the upstream target.

I think I probably need to do my CMake homework here, and am totally happy to revisit this later when I have a better understanding.

Reply:
Not at all.... You have a great point and I am glad we revisit this and change things.

My understating is that `PRIVATE` properties are used internally to build the target. Hence, `PRIVATE` models the `build requirements` of a target.

On the other hand, `INTERFACE` properties are used externally by users of the target. Hence, `INTERFACE` properties models the `usage requirements` of a target.

Reply:
What initially confused me (and possibly what tripped @mattyjams up as well?), was this bit (emphasis mine):

> This is now added in the PRIVATE entry list and **projects that links against mayaUsd now need to explicitly add ${CMAKE_BINARY_DIR}/include in their target_include_directories.**

On re-reading this thread, I now assume that when you talked about external projects including this dir explicitly, you were speaking hypothetically.  Is this correct? Or are there actually cases where an external project would want / need to do this?

Reply:
Correct, my comment was mostly hypothetical. I am happy if we want to revisit this but let's do it after this PR merged.